### PR TITLE
Add learned activations L0 norm metric

### PR DESCRIPTION
**Pull request by lucyfarnik**
Closes issue #87

### CODE REVIEW

**Comment by lucyfarnik** [comment]
Something to note on this PR: the L0NormMetric.create_progress_bar_postfix method isn't implemented yet. The abstract class didn't provide any info on what this method should do so I didn't have anything to go off of here.

**Review by alan-cooney — CHANGES_REQUESTED** [review]
The L0 loss function already logs to wandb, so this doesn't add anything I'm afraid.

However if we decide we want L0 on the reconstructed neurons it would be reasonable to add this.

Btw I've also removed `create_progress_bar_postfix` and renamed `create_weights_and_biases_log` to just `calculate`.

**Comment by lucyfarnik** [comment]
I'm slightly confused, I'm pretty sure the only Lp norm that's getting logged to wandb is the L1? At least this was what I was seeing when running the code before I added these features in:
Screenshot 2023-11-23 at 7 44 32 AM


Re removing `create_progress_bar_postfix`: sounds great! I think that was the right call

File: sparse_autoencoder/metrics/train/l0_norm_metric.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+class L0NormMetric(AbstractTrainMetric):
+    """L0 (sparsity) norm metric."""
+
+    def calculate(self, data: TrainMetricData) -> dict[str, float]:
+        """Create a log item for Weights and Biases."""
+        # The L0 norm is the number of non-zero elements
+        # (We're averaging over the batch)
+        acts = data.learned_activations
+        value = (torch.sum(acts != 0) / acts.size(0)).item()
+        return {"l0_norm": value}
```
Comment (alan-cooney) [inline sparse_autoencoder/metrics/train/l0_norm_metric.py:22]: A few suggestions here:

- The docstring should have an Args: and Returns: section
- The comment, that explains what L0 norm is, should be in the class docstring.
- Better to use descriptive names for variables (so rather than `acts` if you want to create a shorter reference, use `learned_activations`. 
- The resulting value to go to hugging face should have a descriptive name e.g. `learned_activations_mean_l0_norm`
- It's easier to understand what's going on if you split the value calc into it's underlying calcs, e.g.:

```
batch_size = data.learned_activations.size(0)
n_non_zero_activations = torch.count_nonzero(data.learned_activations)
average_l0_norm = n_non_zero_activations / batch_size
```

File: sparse_autoencoder/metrics/train/tests/test_abstract_metric.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+def test_create_weights_and_biases_log(dummy_metric: DummyMetric) -> None:
+    """Test the calculate method."""
+    data = TrainMetricData(
+        input_activations=torch.ones((1, 3)),
+        learned_activations=torch.ones((1, 3)),
+        decoded_activations=torch.ones((1, 3)),
+    )
+    log = dummy_metric.calculate(data)
+    expected = 3.0
+    assert log["dummy_metric"] == expected
```
Comment (alan-cooney) [inline sparse_autoencoder/metrics/train/tests/test_abstract_metric.py:46]: I think we can actually probably just delete this? It doesn't seem to be checking anything the type checking system isn't already checking.

File: sparse_autoencoder/metrics/train/tests/test_l0_norm_metric.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+def test_l0_norm_metric(l0_norm_metric: L0NormMetric) -> None:
+    """Test the L0NormMetric."""
+    learned_activations = torch.tensor([[1.0, 0.0, 0.0], [0.0, 0.01, 2.0]])
+    data = TrainMetricData(
+        input_activations=torch.zeros_like(learned_activations),
+        learned_activations=learned_activations,
+        decoded_activations=torch.zeros_like(learned_activations),
+    )
+    log = l0_norm_metric.calculate(data)
+    expected = 2 / 3
```
Comment (alan-cooney) [inline sparse_autoencoder/metrics/train/tests/test_l0_norm_metric.py:24]: Should be 3/2 right?

**Review by alan-cooney — CHANGES_REQUESTED** [review]
Ignore my previous review - was mixing things up!

I've added some suggestions.

I've also merged in main already and fixed the merge conflicts for you, as that was a result of me refactoring.

**Review by alan-cooney — APPROVED** [review]
